datalad_next/tree.py

```diff
@@ -0,0 +1,1130 @@
+                    **res_dict
+                )
+            else:
+                yield get_status_dict(
+                    status="ok",
+                    **res_dict
+                )
+
+    @staticmethod
+    def custom_result_renderer(res, **kwargs):
+        """
+        Each node is printed on one line. The string uses the format:
+        ``[<indentation>] [<branch_tip_symbol>] <path> [<ds_marker]``
+
+        Example line:
+        ``│   │   ├── path_dir_level3``
+        """
+        from datalad.support import ansi_colors
+
+        # get values from result record
+        node_type = res["type"]
+        node_path = res["path"]
+        depth = res["depth"]
+        exhausted_levels = res["exhausted_levels"]
+
+        # build indentation string
+        indentation = ""
+        if depth > 0:
+            indentation_symbols_for_levels = [
+                ("│"
+                 if level not in exhausted_levels
+                 else " ") + "   "
+                for level in range(1, depth)
+            ]
+            indentation = "".join(indentation_symbols_for_levels)
+
+        # build prefix (tree branch tip)
+        prefix = ""
+        if depth > 0:  # root node has no prefix
+            is_last_child = depth in exhausted_levels
+            prefix = "└──" if is_last_child else "├──"
+
+        # build dataset marker if dataset
+        ds_marker = ""
+        if node_type == "dataset":
+            ds_absolute_depth = res["dataset_abs_depth"]
+            ds_is_installed = res["dataset_is_installed"]
+
+            ds_marker_depth = ansi_colors.color_word(
+                f"DS~{ds_absolute_depth}",
+                ansi_colors.WHITE)
+            install_flag = " (not installed)" if not ds_is_installed else ""
+            ds_marker = f"[{ds_marker_depth}]" + install_flag
+
+        # build path string with optional color
+        # display only root directory with full path, all other nodes
+        # with basename
+        path = node_path if depth == 0 else Path(node_path).name
+        color_for_type = {
+            "dataset": ansi_colors.MAGENTA,
+            "directory": ansi_colors.BLUE,
+            "file": None,
+            "broken_symlink": ansi_colors.RED
+        }
+        # ANSI color for the path, if terminal colors are enabled
+        color = color_for_type[node_type]
+        if color is not None:
+            path = ansi_colors.color_word(path, color)
+        if res.get("is_broken_symlink", False):
+            path = ansi_colors.color_word(path,
+                                          color_for_type["broken_symlink"])
+
+        # set suffix for directories
+        dir_suffix = ""
+        if depth > 0 and node_type in ("directory", "dataset"):
+            dir_suffix = "/"
+
+        # append symlink target if symlink
+        symlink_target = ""
+        if "symlink_target" in res:
+            symlink_target = " -> " + res["symlink_target"]
```

Comment:
Added rendering of symlinks with typical UNIX formatting `<link> -> <target>`, as in `tree` and `ls`.